Add delete_account method to FxAPom

bobsilverberg opened the pull request:
Also, update tests to verify that created account can be used for logging in

This fixes issue #4 

@davehunt r?

I realize that the module(s) have issues that you'd still like to address, but I think we can add this feature now and still refactor later.

davehunt commented on conftest.py:
```diff
@@ -13,3 +16,25 @@ def pytest_funcarg__mozwebqa(request):
+
+
+@pytest.fixture
+def dev_account(request):
```
We could parameterize this fixture so the tests using it will run against multiple instances. See http://pytest.org/latest/fixture.html#parametrizing-a-fixture for details.

bobsilverberg replied:
Thanks. I changed that, although we still need a `dev_account` fixture as we don't need or want all of the tests running against prod.

davehunt commented on conftest.py:
```diff
@@ -13,3 +16,25 @@ def pytest_funcarg__mozwebqa(request):
+
+@pytest.fixture
+def dev_account(request):
+    acct = FxATestAccount(DEV_URL).create_account()
```
Could we call this `account`?

davehunt commented on fxapom/fxapom.py:
```diff
@@ -54,6 +55,16 @@ def create_account(self):
         return self
 
+    def delete_account(self):
+        if not self.deleted:
```
Why do we need this check? What happens if we try to delete an account twice?

bobsilverberg replied:
We get an exception if we try to delete an account that doesn't exist. I have reworked the code to catch this exception.

davehunt commented on tests/test_fxa_test_account.py:
```diff
@@ -0,0 +1,46 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import re
```
Nit: Add a blank line after license.

davehunt commented on tests/test_fxa_test_account.py:
```diff
@@ -0,0 +1,46 @@
+        try:
+            acct.login()
+        except ClientError:
+            raise AssertionError('Login unsuccessful. Account not properly created/verified.')
```
This will conceal the ClientError, which might make investigation difficult if there are failures. As the unexpected exception will cause the test to error, can we just leave it?

bobsilverberg replied:
Sure, we can do that.

davehunt commented on tests/test_fxa_test_account.py:
```diff
@@ -0,0 +1,46 @@
+
+    def test_new_account_pw_does_not_contain_numbers(self, dev_account):
+        test_regex = re.compile('\d')
+        assert test_regex.search(dev_account.password) is None
```
I don't think we really need to compile the regex here, we can just use [re.search](https://docs.python.org/2/library/re.html#re.search)

bobsilverberg replied:
Correct. Fixed.

davehunt commented on tests/test_fxa_test_account.py:
```diff
@@ -0,0 +1,46 @@
+            dev_account.login()
+        except ClientError:
+            return
+        raise AssertionError('Login successful. Account was not deleted.')
```
Same comment on masking the exception. If the `ClientError` is not helpful as it is then we should consider catching in fxapom and raising a more useful exception.

bobsilverberg replied:
I have reworked this as you suggested, catching the specific error we care about in `login` and throwing a custom exception.

davehunt commented:
I think my issue is that our `FxATestAccount` object is not in fact a test account - it's more of a test account manager. Maybe a simple rename would help with that, but I also wonder if it could represent an account (and create the account on instantiation). I also think we should implement a [__del__](https://docs.python.org/2/reference/datamodel.html#object.__del__) method to delete the account when the object is destroyed - this could either call the delete method you've created here, or could mean we don't need to explicitly delete the account.

bobsilverberg commented:
@davehunt I have addressed your comments and the code seems much better for it. I agree with your idea of creating an account on instantiation and deleting it via `__del__`, but I'm wondering if that should be in a separate PR. It may be I try to address that before you return, in which case I'll add it here, but if I don't please review this one as is.

davehunt commented on fxapom/fxapom.py:
```diff
@@ -54,6 +58,29 @@ def create_account(self):
+            # cause us to throw a custom exception
+            # https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#response-format
+            if err.errno == 102:
+                raise AccountNotFoundException('FxA Account Not Found')
```
Out of curiosity, what does the ClientError message look like? I assume it's not particularly friendly, hence the need for our custom exception?

bobsilverberg replied:
The `ClientError` looks like:

`ClientError: Unknown account`

The issue is that there can be many different types of exceptions that are all `ClientError`, so simply looking for that exception type in the test wouldn't work. I thought it better to catch `ClientError` in login, and if it is the specific exception that we care about (which is `errno` 102) then to raise my own custom exception, which I can then catch in the test. The test in question is `test_delete_account` and it wants `login` to fail with that specific exception.

I thought this better than adding the logic that looks for `errno == 102` into the test itself.
Does that make sense?

davehunt replied:
It does make sense, and I don't mind custom exceptions when there's a good reason for them. :)

davehunt commented on tests/test_fxa_test_account.py:
```diff
@@ -0,0 +1,33 @@
+            account.login()
+        except AccountNotFoundException:
+            return
+        raise AssertionError('Login successful. Account was not deleted.')
```
Can we just use pytest.raises here: http://pytest.org/latest/assert.html#assertions-about-expected-exceptions

davehunt commented:
This is a good enhancement, I don't think we need to wait for the `__init__` and `__del__` changes, that can come later.

davehunt commented:
r+ with my comments addressed

bobsilverberg commented:
Thanks @davehunt. I changed the code to use `pytest.raises()` which is much better, but I also believe requires my custom exception that you were concerned about. Because of that, and your r+wc, I'm going to merge this. If you still have issues with the custom exception I'd be happy to hear about them.

stephendonner commented:
:+1:
